File: frontend/src/lib/matching/incomeBands.ts
/**
 * India household income-band classification (PRICE ICE 360°).
 *
 * Source of truth (NOT a Government of India statutory definition):
 * PRICE ICE 360° / "The Rise of India's Middle Class" — 2020–21 prices.
 * Primary PDF: https://www.price360.in/Executive_Summary_Middle_Class.pdf
 *
 * ₹15 lakh is the Seekers → Strivers cut (Strivers are ₹15–30 lakh annual).
 * These bands are for UX labels and soft ranking only — NEVER invent scheme
 * eligibility ceilings (max_annual_income) from PRICE bands.
 * INDIA-ONLY: never map USD (or other currencies) onto these INR thresholds;
 * non-India countries must return null band fields and skip PRICE boosts.
 *
 * Mirror of backend/app/income_bands.py — keep thresholds/labels in sync.
 */

export const INCOME_BAND_SOURCE =
  "PRICE ICE 360° (2020–21 prices; not official GoI)";

export const INCOME_BAND_NOTE =
  "India-only household income bands from PRICE ICE 360° / The Rise of India's Middle Class " +
  "(2020–21 prices). Not a Government of India statutory classification. " +
  "Used for UX labels and soft ranking only — does not change scheme eligibility rules. " +
  "₹15 lakh is the Seekers→Strivers boundary (Strivers: ₹15–30 lakh). " +
  "Do not apply these INR bands to United States or other countries.";

export interface IncomeBandResult {
  band_id: string | null;
  band_label: string | null;
  income_class: string | null;
  source: string | null;
  note: string | null;
}

const BANDS: Array<{
  lo: number;
  hi: number | null;
  band_id: string;
  band_label: string;
}> = [
  { lo: 0, hi: 124_999, band_id: "destitute", band_label: "Destitute" },
  { lo: 125_000, hi: 499_999, band_id: "aspirer", band_label: "Aspirer" },
  {
    lo: 500_000,
    hi: 1_499_999,
    band_id: "seeker",
    band_label: "Middle class (Seekers)",
  },
  {
    lo: 1_500_000,
    hi: 2_999_999,
    band_id: "striver",
    band_label: "Upper middle class (Strivers)",
  },
  { lo: 3_000_000, hi: 4_999_999, band_id: "near_rich", band_label: "Near rich" },
  { lo: 5_000_000, hi: 9_999_999, band_id: "clear_rich", band_label: "Clear rich" },
  {
    lo: 10_000_000,
    hi: 19_999_999,
    band_id: "sheer_rich",
    band_label: "Sheer rich",
  },
  { lo: 20_000_000, hi: null, band_id: "super_rich", band_label: "Super rich" },
];

const BAND_TO_CLASS: Record<string, string> = {
  destitute: "destitute",
  aspirer: "aspirer",
  seeker: "middle_class",
  striver: "middle_class",
  near_rich: "rich",
  clear_rich: "rich",
  sheer_rich: "rich",
  super_rich: "rich",
};

export const BAND_RANGE_HINT: Record<string, string> = {
  destitute: "under ₹1.25 lakh",
  aspirer: "₹1.25–5 lakh",
  seeker: "₹5–15 lakh",
  striver: "₹15–30 lakh",
  near_rich: "₹30–50 lakh",
  clear_rich: "₹50 lakh–1 crore",
  sheer_rich: "₹1–2 crore",
  super_rich: "₹2 crore+",
};

function normCountry(country: string | null | undefined): string {
  if (!country) return "india";
  return String(country)
    .trim()
    .toLowerCase()
    .replace(/-/g, "_")
    .replace(/ /g, "_");
}

export function isIndiaCountry(country: string | null | undefined): boolean {
  return normCountry(country || "India") === "india";
}

const EMPTY: IncomeBandResult = {
  band_id: null,
  band_label: null,
  income_class: null,
  source: null,
  note: null,
};

export function classifyIndiaAnnualIncome(
  annual: number | null | undefined,
  country: string | null | undefined = "India",
): IncomeBandResult {
  // Hard country gate — PRICE INR thresholds must never run for US/others.
  if (!isIndiaCountry(country)) return { ...EMPTY };
  if (annual == null) return { ...EMPTY };
  const value = Number(annual);
  if (!Number.isFinite(value) || value < 0) return { ...EMPTY };

  for (const b of BANDS) {
    if (value < b.lo) continue;
    if (b.hi == null || value <= b.hi) {
      return {
        band_id: b.band_id,
        band_label: b.band_label,
        income_class: BAND_TO_CLASS[b.band_id],
        source: INCOME_BAND_SOURCE,
        note: INCOME_BAND_NOTE,
      };
    }
  }
  return {
    band_id: "super_rich",
    band_label: "Super rich",
    income_class: "rich",
    source: INCOME_BAND_SOURCE,
    note: INCOME_BAND_NOTE,
  };
}

export function formatBandSentence(band: IncomeBandResult): string | null {
  if (!band.band_id || !band.band_label) return null;
  const hint = BAND_RANGE_HINT[band.band_id] || "";
  const rangePart = hint ? ` (${hint})` : "";
  return (
    `Based on PRICE ICE 360 household bands (2020–21 prices): ` +
    `${band.band_label}${rangePart}. ` +
    `Not an official government classification — does not change scheme rules; ` +
    `only helps sort and explain results.`
  );
}

/** Soft ranking boosts — mirror backend/app/income_bands.py */
export const MIDDLE_HIGH_BOOST_TAGS: Record<string, number> = {
  "upper-middle-class": 0.1,
  "high-income-eligible": 0.1,
  "middle-class": 0.08,
  tax: 0.06,
  savings: 0.06,
  universal: 0.05,
};
export const MIDDLE_HIGH_BOOST_CAP = 0.12;
export const MIDDLE_HIGH_BANDS = new Set([
  "seeker",
  "striver",
  "near_rich",
  "clear_rich",
  "sheer_rich",
  "super_rich",
]);

export const LOW_INCOME_BOOST_TAGS: Record<string, number> = {
  welfare: 0.04,
  bpl: 0.04,
  pension: 0.03,
};
export const LOW_INCOME_IMPLIES_BOOST = 0.04;
export const LOW_INCOME_BOOST_CAP = 0.05;
export const LOW_INCOME_BANDS = new Set(["destitute", "aspirer"]);

export function incomeBandScoreBoost(
  bandId: string | null | undefined,
  tags: string[] | null | undefined,
  impliesLowIncome = false,
): number {
  if (!bandId) return 0;
  const tagSet = new Set(
    (tags || []).map((t) => String(t).trim().toLowerCase()).filter(Boolean),
  );

  if (MIDDLE_HIGH_BANDS.has(bandId)) {
    let bump = 0;
    for (const [tag, amount] of Object.entries(MIDDLE_HIGH_BOOST_TAGS)) {
      if (tagSet.has(tag)) bump = Math.max(bump, amount);
    }
    return Math.min(bump, MIDDLE_HIGH_BOOST_CAP);
  }

  if (LOW_INCOME_BANDS.has(bandId)) {
    let bump = 0;
    for (const [tag, amount] of Object.entries(LOW_INCOME_BOOST_TAGS)) {
      if (tagSet.has(tag)) bump = Math.max(bump, amount);
    }
    if (impliesLowIncome) bump = Math.max(bump, LOW_INCOME_IMPLIES_BOOST);
    return Math.min(bump, LOW_INCOME_BOOST_CAP);
  }

  return 0;
}

File: backend/app/income_bands.py
"""India household income-band classification (PRICE ICE 360°).

Source of truth (NOT a Government of India statutory definition):
PRICE ICE 360° / "The Rise of India's Middle Class" — 2020–21 prices.
Primary PDF: https://www.price360.in/Executive_Summary_Middle_Class.pdf

₹15 lakh is the Seekers → Strivers cut (Strivers are ₹15–30 lakh annual).
These bands are for UX labels and soft ranking only — NEVER invent scheme
eligibility ceilings (max_annual_income) from PRICE bands.
"""

from __future__ import annotations

from typing import Any, TypedDict

# Short citation for API / UI.
INCOME_BAND_SOURCE = "PRICE ICE 360° (2020–21 prices; not official GoI)"

INCOME_BAND_NOTE = (
    "Household income bands from PRICE ICE 360° / The Rise of India's Middle Class "
    "(2020–21 prices). Not a Government of India statutory classification. "
    "Used for UX labels and soft ranking only — does not change scheme eligibility rules. "
    "₹15 lakh is the Seekers→Strivers boundary (Strivers: ₹15–30 lakh)."
)

# (min_inclusive, max_inclusive_or_None, band_id, band_label)
# Upper bound None means open-ended (≥).
_BANDS: list[tuple[float, float | None, str, str]] = [
    (0, 124_999, "destitute", "Destitute"),
    (125_000, 499_999, "aspirer", "Aspirer"),
    (500_000, 1_499_999, "seeker", "Middle class (Seekers)"),
    (1_500_000, 2_999_999, "striver", "Upper middle class (Strivers)"),
    (3_000_000, 4_999_999, "near_rich", "Near rich"),
    (5_000_000, 9_999_999, "clear_rich", "Clear rich"),
    (10_000_000, 19_999_999, "sheer_rich", "Sheer rich"),
    (20_000_000, None, "super_rich", "Super rich"),
]

# Coarser income_class: destitute | aspirer | middle_class (seeker+striver) | rich (≥30L)
_BAND_TO_CLASS: dict[str, str] = {
    "destitute": "destitute",
    "aspirer": "aspirer",
    "seeker": "middle_class",
    "striver": "middle_class",
    "near_rich": "rich",
    "clear_rich": "rich",
    "sheer_rich": "rich",
    "super_rich": "rich",
}

# Human-readable range hints for UI (lakh-style).
BAND_RANGE_HINT: dict[str, str] = {
    "destitute": "under ₹1.25 lakh",
    "aspirer": "₹1.25–5 lakh",
    "seeker": "₹5–15 lakh",
    "striver": "₹15–30 lakh",
    "near_rich": "₹30–50 lakh",
    "clear_rich": "₹50 lakh–1 crore",
    "sheer_rich": "₹1–2 crore",
    "super_rich": "₹2 crore+",
}


class IncomeBandResult(TypedDict):
    band_id: str | None
    band_label: str | None
    income_class: str | None
    source: str | None
    note: str | None


def _norm_country(country: str | None) -> str:
    if not country:
        return "india"
    return str(country).strip().lower().replace("-", "_").replace(" ", "_")


def is_india_country(country: str | None) -> bool:
    return _norm_country(country or "India") in {"india"}


def classify_india_annual_income(
    annual: float | None,
    *,
    country: str | None = "India",
) -> IncomeBandResult:
    """Classify annual household income into PRICE ICE 360° bands.

    Returns null band fields when annual is None, or when country is not India
    (US / other stubs deferred — skip non-India bands this PR).
    """
    empty: IncomeBandResult = {
        "band_id": None,
        "band_label": None,
        "income_class": None,
        "source": None,
        "note": None,
    }
    if not is_india_country(country):
        return empty
    if annual is None:
        return empty
    try:
        value = float(annual)
    except (TypeError, ValueError):
        return empty
    if value < 0:
        return empty

    for lo, hi, band_id, band_label in _BANDS:
        if value < lo:
            continue
        if hi is None or value <= hi:
            return {
                "band_id": band_id,
                "band_label": band_label,
                "income_class": _BAND_TO_CLASS[band_id],
                "source": INCOME_BAND_SOURCE,
                "note": INCOME_BAND_NOTE,
            }
    # Fallback (should not hit — super_rich is open-ended)
    return {
        "band_id": "super_rich",
        "band_label": "Super rich",
        "income_class": "rich",
        "source": INCOME_BAND_SOURCE,
        "note": INCOME_BAND_NOTE,
    }


def format_band_sentence(band: IncomeBandResult | dict[str, Any]) -> str | None:
    """One plain UX sentence; None if band unknown."""
    band_id = band.get("band_id")
    band_label = band.get("band_label")
    if not band_id or not band_label:
        return None
    hint = BAND_RANGE_HINT.get(str(band_id), "")
    range_part = f" ({hint})" if hint else ""
    return (
        f"Based on PRICE ICE 360 household bands (2020–21 prices): "
        f"{band_label}{range_part}. "
        f"Not an official government classification — does not change scheme rules; "
        f"only helps sort and explain results."
    )


# ---------------------------------------------------------------------------
# Soft ranking boosts (deterministic; secondary to hard eligibility)
# ---------------------------------------------------------------------------

# For seeker / striver / *_rich India profiles: modest bump for middle/tax/savings tags.
# Take the MAX of matching tag boosts, then cap at MIDDLE_HIGH_BOOST_CAP.
MIDDLE_HIGH_BOOST_TAGS: dict[str, float] = {
    "upper-middle-class": 0.10,
    "high-income-eligible": 0.10,
    "middle-class": 0.08,
    "tax": 0.06,
    "savings": 0.06,
    "universal": 0.05,
}
MIDDLE_HIGH_BOOST_CAP = 0.12
MIDDLE_HIGH_BANDS = frozenset(
    {"seeker", "striver", "near_rich", "clear_rich", "sheer_rich", "super_rich"}
)

# For destitute / aspirer: tiny optional boost for welfare / implies_low schemes.
LOW_INCOME_BOOST_TAGS: dict[str, float] = {
    "welfare": 0.04,
    "bpl": 0.04,
    "pension": 0.03,
}
LOW_INCOME_IMPLIES_BOOST = 0.04
LOW_INCOME_BOOST_CAP = 0.05
LOW_INCOME_BANDS = frozenset({"destitute", "aspirer"})


def income_band_score_boost(
    band_id: str | None,
    tags: list[str] | None,
    *,
    implies_low_income: bool = False,
) -> float:
    """Return a small deterministic score bump for ranking (0 if none).

    Does NOT change hard gates (max_annual_income, implies_low_income exclude).
    Applied after base eligibility score; sort still prefers likely_eligible first.
    """
    if not band_id:
        return 0.0
    tag_set = {str(t).strip().lower() for t in (tags or []) if t}

    if band_id in MIDDLE_HIGH_BANDS:
        bump = 0.0
        for tag, amount in MIDDLE_HIGH_BOOST_TAGS.items():
            if tag in tag_set:
                bump = max(bump, amount)
        return min(bump, MIDDLE_HIGH_BOOST_CAP)

    if band_id in LOW_INCOME_BANDS:
        bump = 0.0
        for tag, amount in LOW_INCOME_BOOST_TAGS.items():
            if tag in tag_set:
                bump = max(bump, amount)
        if implies_low_income:
            bump = max(bump, LOW_INCOME_IMPLIES_BOOST)
        return min(bump, LOW_INCOME_BOOST_CAP)

    return 0.0
